fs-storage/src/folder_storage.rs

```
@@ -0,0 +1,784 @@
+    ///
+    /// Note: if the folder storage already exists, the data will be read from the folder
+    /// without overwriting it.
+    pub fn new(label: String, path: &Path) -> Result<Self> {
+        let mut storage = Self {
+            label,
+            path: PathBuf::from(path),
+            modified: BTreeMap::new(),
+            written_to_disk: BTreeMap::new(),
+            data: BTreeMap::new(),
+            deleted_keys: BTreeSet::new(),
+        };
+
+        if Path::exists(path) {
+            storage.read_fs()?;
+        }
+
+        Ok(storage)
+    }
+
+    /// Load mapping from folder storage
+    fn load_fs_data(&mut self) -> Result<()> {
+        if !self.path.exists() {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Folder does not exist".to_owned(),
+            ));
+        }
+
+        if !self.path.is_dir() {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Path is not a directory".to_owned(),
+            ));
+        }
+
+        let mut data = BTreeMap::new();
+
+        for entry in fs::read_dir(&self.path)? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_file()
+                && path
+                    .extension()
+                    .map_or(false, |ext| ext == "json")
+            {
+                let key: K = self.extract_key_from_file_path(&path)?;
+                let file = File::open(&path)?;
+                let value: V =
+                    serde_json::from_reader(file).map_err(|err| {
+                        ArklibError::Storage(
+                            self.label.clone(),
+                            err.to_string(),
+                        )
+                    })?;
+
+                data.insert(key.clone(), value);
+
+                if let Ok(metadata) = fs::metadata(&path) {
+                    if let Ok(modified) = metadata.modified() {
```

We should return an error here if we can't retrieve it from disk.